ncm-filesystems/src/main/perl/filesystems.pm

```diff
@@ -1,116 +1,143 @@
+use Fcntl qw(:seek);
+use Readonly;
+use Cwd qw(abs_path);
 
-# Returns a hash with the canonical paths to the protected mountpoints
-# as its keys. I This small overhead may simplify the code later.
-sub protected_hash
-{
-	my $config = shift;
-	my %ph;
+Readonly my $UMOUNT => "/bin/umount";
+Readonly my $FSTAB_CMP_PATH => '/software/components/fstab';
+Readonly my $FS_TREE => '/system/filesystems';
+use parent qw(NCM::Component::fstab);
 
-	my $pl = $config->getElement (PROTECTED_PATH)->getTree;
-
-	foreach my $i (@$pl) {
-		$ph{$i} = 0;
-		my $path = abs_path("$i/");
-		$ph{$path} = 0 if $path;
-	}
-	return %ph;
-}
+our $EC = LC::Exception::Context->new()->will_store_all();
+our $NoActionSupported = 0; # First needs check/fixing in ncm-lib-blockdevices issue #74
 
 # Returns a hash with the mountpoints of the filesystems defined on
 # the profile as its keys, and the filesystem objects as its values.
 sub fshash
 {
-	my $fsl = shift;
-	my %fsh;
+    my ($self, $fsl) = @_;
+    my %fsh;
 
-	$fsh{$_->{mountpoint}} = $_ foreach @$fsl;
-	return %fsh;
+    $fsh{$_->{mountpoint}} = $_ foreach @$fsl;
+    return %fsh;
 }
 
-# Frees space on the system. It removes filesystems that are not
-# present in the profile anymore, and are *not* present in the
-# protected hash either. Returns 0 on success, -1 on error.
-sub free_space
+#  Removes filesystem and deletes stuff that is not present in the %mounts 
+#  argument from fstab.
+sub delete_outdated
 {
-	my ($self, $cfg, $protected, @fs) = @_;
-
-	my %ph = %$protected;
-	my %fsh = fshash (\@fs);
-
-	my $fl = output ("grep", "^[[:space:]]*#", "/etc/fstab", "-v");
-
-	$self->info ("Checking filesystems not defined in the profile");
-	my @fstab = split ("\n+", $fl);
-	foreach my $l (@fstab) {
-		$self->debug (5,"Fstab line: $l");
-		$l =~ m{^\S+\s+(\S+)\s} or next;
-		my $mount = $1;
-		next if(exists $fsh{$mount} || exists $ph{$mount}
-			    || exists($ph{abs_path($mount)}));
-		$self->verbose("Removing filesystem $mount");
-		my $f = NCM::Filesystem->new_from_fstab ($l, log => $self);
-		$f->remove_if_needed==0 or return -1;
-
-	}
-	return 0;
+    my ($self, $fstab, $mountsr, $config, $remove) = @_;
+
+    my @rm;
+    my %mounts = %$mountsr;
+    seek($fstab, 0, SEEK_SET);
+
+    while (my $f = <$fstab>) {
+    	my $mount = $self->mount_from_entry($f) or next;
+    	if (!exists ($mounts{$mount}) && !exists($mounts{abs_path($mount)})) {
+    	    CAF::Process->new ([$UMOUNT, $mount],
+    			       log => $self)->run();
+    	    push (@rm, $f);
+            $self->verbose ("Scheduling for removal of fstab: $mount");
+            $self->verbose ("Scheduling for removal filesystem $mount") if $remove;
+    	}
+    }
+
+    foreach my $outdated (@rm) {
+        if ($remove) {
+            my $fsrm = NCM::Filesystem->new_from_fstab ($outdated, $config, log => $self);
+            $self->info("Removing filesystem for $fsrm->{mountpoint}");
+            $fsrm->remove_if_needed==0 or return 0;
```

Comment:
i mean it's ver very cryptic. especially if ncm-lib-blockdevices does not report an error. it would be better to explcitly,

```perl
if ($fsrm->remove_if_needed) {
    $self->error("error $fsrm->{mountpoint}");
} else {
    $self->info("removed $fsrm->{mountpoint}")
}
```

you are after all deteing a filesystem, it's nice we have this reporter/logged